<p style="text-align: center;"><font size="8"><b>Functions and Modules</b></font><br>


# Calling Functions

Pure functions exist as methods that can be called outside the context of a particular class. For example we're already seen the `round` and `len` functions. Remember that we call `round(a)`, not `a.round()`. Python provides many built-in functions. 

![functions](https://github.com/lukasbystricky/ISC-3313/blob/master/lectures/chapter2/images/functions.png?raw=true)

In [1]:
pow(2,3)

8

In [2]:
min(-2,-2.4,9.0)

-2.4

In [3]:
ord("a")

97

# Print function

After performing a computation we need to see the result. In an interactive or notebook session we can simply type the name of the variable to see what it is.

In [4]:
a = 1
a

1

When executing scripts however, we can no longer to this. Instead we use the `print` command (in fact we've already used this several times).

The `print` function is used to print information to the console. 

In [5]:
print(a)
print("a")

1
a


Note the syntax used in the textbook, `print a`, (no parentheses) does not work in Python 3. However the syntax above works in both Python 2 and Python 3.


Of course, `print` can be used to display more useful information. For example, suppose we have a variable called `t` that represents some length of time in seconds. We can use `print` to display not only `t`, but also the units by calling `print` with multiple arguments.

In [6]:
t = 5.6
print(t, "s")

5.6 s


The `print` function automatically inserts a space between the two arguments. If we wish the avoid this, we can combine the two arguments into 1.

In [7]:
print(str(t)+"s")

5.6s


Note that we have to convert `round(t,1)` to a string first before "adding" (the correct term is *concatenating*) it to `s`. The command

In [8]:
print(t+"s")

TypeError: unsupported operand type(s) for +: 'float' and 'str'

is illegal because we are attempting to add a `float` to a `str`.

# Modules

All the functions listed above are built-in to Python. This means they are automatically avaliable once we start Python. But you'll notice that there's not actually all that many built-in functions. What if we want to take the logarithm of a number for example?

There are hundreds of other useful functions and classes that have been developped for Python which are not automatically loaded, but instead placed into specialized libraries called *modules* that can be individually loaded as needed.

If you installed Python using Anaconda, you will already have many modules installed. Today we will look at the math module, the SymPy module and the NumPy modules, all of which come with Anaconda. 

## Math module

The math module provides functions to do mathematical operations beyond addition, multiplication, exponentiation etc. For example suppose we want to take the cosine of a number. This is not built-in to Python, however a function called `cos` in the math library does this. 

To use the `cos` function we must import it from the math module. There are three possible ways to do this. Whichever method you use should be done at the beginning of your code (this is not strictly speaking necessary, but it is the most common placement, at the very least you have to do this *before* you use the function).

1) Import the entire module. 

In [2]:
import math

At this point we still cannot use the `cos` command directly, we must specifically tell Python where the function is coming from using a *qualified name*.

In [3]:
cos(2)

NameError: name 'cos' is not defined

In [4]:
math.cos(2)

-0.4161468365471424

2) Specifically importing `cos` from the math library. If we are using `cos` several times in the code, this can avoid repeated typing.

In [5]:
from math import cos
cos(2)

-0.4161468365471424

3) Import everything in the module by using the \* wildcard. This can be an attractive option but is generally discouraged because different modules may use the same name for different functions. This method of importing imports not only the `cos` function, but all functions from the math library, for example `sqrt` and `tan`.

In [6]:
from math import *
cos(2)

-0.4161468365471424

In [7]:
sqrt(2)

1.4142135623730951

# Expressions

Before looking at other modules, let's look quickly at expressions. We have already seen several expressions in isolation (e.g. 18+5.5).

It is quite common the perform several operations as part of a single expression.

In [15]:
a = 18 + 5.5 + 1
a

24.5

In this case behind the scenes Python adds 18 and 5.5 to get 23.5, it then adds 1 to get 24.5. In this case the order of the two operations does not matter, however in more complicated expressions the order can be important.

In [16]:
a = 18*9**2/4
a

364.5

In [17]:
a = 9/4**2*18
a

10.125

## Precedence

When there are two or more operations as part of an expression, we must figure out some way to determine which operation is performed first. We say that an operation that is performed fist is given *precedence* over the others.

Mathematical expressions in Python follow standard algebraic conventions:
1. Brackets
2. Exponents
3. Division/Multiplication
4. Addition/Subtraction

For example in the expression `1 + 2*3` the multiplication is done first, followed by the addition. 

In Python, as in algebra we can use brackets to prioritize an operation.

In [18]:
1+2*3

7

In [19]:
(1+2)*3

9

Most operations with equal precedence are evaluated left to right, again to mimic standard algebraic rules.  

One exception is exponenents which are evaluated right to left, which again is how we typically think of exponents.

$4^{3^2} = 4^9$

In [20]:
4**3**2

262144

In [21]:
4**(3**2)

262144

In [22]:
(4**3)**2

4096

Even though precedence rules are based on algebraic rules, they are enforced for any data type. 

In [23]:
"a"*3+"b"

'aaab'

### Example

When boiling an egg, it has been determined that the time it takes for the center of the yolk to reach a desired temperature $T$ is given by
$$ t = \frac{M^{2/3}c\rho^{1/3}}{K\pi^2(4\pi/3)^{2/3}}\ln\left(0.76\frac{T_0 - 100}{T - 100}\right)$$

where
* $M$ is the mass of the egg
* $\rho$ is the density
* $c$ is the specific heat capacity
* $K$ is the thermal conductivity
* $T_0$ is the temperature at $t=0$

$$ t = \frac{M^{2/3}c\rho^{1/3}}{K\pi^2(4\pi/3)^{2/3}}\ln\left(0.76\frac{T_0 - 100}{T - 100}\right)$$

In [24]:
T = 70 # desired temperature
M = 47
rho = 1.038
c = 3.7
K = 5.4e-3
T0 = 4

# compute time according to above formula, we need the natural logarithm function and the pi constant from the math module
from math import log, pi

t = (M**(2/3)*c*rho**(1/3))/(K*pi**2*(4*pi/3)**(2/3))*log(0.76*(T0 - 100)/(T - 100))
print(t)

313.09454902221637


## Exercise 

A quadratic equation can be written as:
$$ ax^2 + bx + c = 0.$$

In general this equation has two (possibly equal) solutions and they are given by the formulas:
\begin{align*}
    x_1 = \frac{-b + \sqrt{b^2 - 4ac}}{2a}\\
    x_2 = \frac{-b - \sqrt{b^2 - 4ac}}{2a}\\
\end{align*}

Use these formulas to compute the solutions of the equation $8x^2 + 16x + 4 = 0$.


In [ ]:
a = ...
b = ...
c = ...

x1 = ...
x2 = ...

## Calling Functions from Within Expressions

Function calls have high precedence. When multiple finction calls are used in the same expression they are typically evaluated from left to right.

In [27]:
person = "George Washington"
person.split()[1]

'Washington'

More complicated expressions are evaluated by first resolving commands inside parentheses.

In [28]:
groceries = ["cereal", "milk", "apple"]
groceries.insert(groceries.index("milk") + 1, "eggs")
groceries

['cereal', 'milk', 'eggs', 'apple']

Here we first must evaluate `groceries.index("milk")` and then add 1 to it to find the index where we wish to insert "eggs".

## Exercise

Write a one line expression to capitalize the word to the immediate right of "milk" in the given list. Assume that you don't know the index of "milk" beforehand.

In [57]:
groceries = ["cereal", "milk", "apple"]

# change milk to Milk
groceries[groceries.index("milk") + 1] = ...

# print new list
print(groceries)

# SymPy

Another potentially useful module is the `SymPy` module (http://www.sympy.org/en/index.html) for symbolic computing. Recall from earlier that computers typically only store 16 digits of a real number. Sometimes this is not enough and we want an exact representation. Alternatively we may have a complictated formula that we want to simplify. In either case, we can use SymPy to manipulate and evaluate analytic formulas. 

As with the `math` module, we must first import the SymPy module. We will import the entire module, however to reduce typing instead of calling it `sympy`, we will call it `sp`.

In [9]:
import sympy as sp

Consider the following example. Say we wanted to evaluate the square root of 2. We already have several ways of doing this:
* `math.sqrt(2)`
* `pow(2,0.5)`
* `2**0.5`

None of these expressions are not exact however. 

In [10]:
a = pow(2,0.5)
print(a*a)

2.0000000000000004


Using Sympy we can represent $\sqrt{2}$ exactly as an object.

In [11]:
b = sp.sqrt(2)
type(b)

sympy.core.power.Pow

In [12]:
print(b)

sqrt(2)


Note that `sp.sqrt` is not the same as `math.sqrt`. Functions that appear in different modules but have the same name is very common. This is why it is discouraged to use the wildcard * to import entire modules.

The real power of SymPy however, comes in the simplification of expressions. Recall that for example we can rewrite $\sqrt{8}$ as $2\sqrt{2}$. Well SymPy does this for us automatically.

In [13]:
print(sp.sqrt(8))

2*sqrt(2)


What's more, say we actually do want more than 16 digits of an irrational number. We can use SymPy's `evalf` function to get an arbitrary number of digits.

In [14]:
a = sp.pi
b = a.evalf(50)
print(b) # print the first 50 digits of pi

3.1415926535897932384626433832795028841971693993751


Much like with the square root function, SymPy also provides us with many other mathematical functions, sine being one of them.

In [15]:
print(sp.sin(a))
print(sp.sin(b))

0
-1.0106957375356833111703700717052567514058479739458e-51


It's important to remember however that this number is not actually stored as a Python `float`.

In [16]:
type(b)

sympy.core.numbers.Float

It's a SymPy float, meaning that as long as we use it in SymPy it will have 50 digits, but if we try to use it in a regular Python expression it will again be converted to a 16 digit representation. 

In [17]:
import math
print(math.sin(b))

1.2246467991473532e-16


## Symbols

So we can use SymPy to manipulate irrational numbers. 

We can also use it to manipulate and simplify expressions containing variables. In SymPy expressions variable are defined using the `symbols` constructor. 

In [28]:
x = sp.symbols("x")
type(x)

sympy.core.symbol.Symbol

Now we can use this variable `x` in an expression. Let's create the expression $8x^2 + 2x + 3$.

In [29]:
expr = 8*x**2 + 2*x + 3
print(expr)

8*x**2 + 2*x + 3


The `expr` variable can now be treated exactly as if it were an equation you have written down on paper. 

We can multiply it by $x$ for example and divide it by $x^2 + 3x + 2x$.

In [30]:
expr = x*expr/(x**2 + 3*x + 2*x)
print(expr)

x*(8*x**2 + 2*x + 3)/(x**2 + 5*x)


SymPy can also be used to do things like simplication of expressions.

In [31]:
print(sp.simplify(expr))

(8*x**2 + 2*x + 3)/(x + 5)


What about differentiation? Suppose we need to compute the derivative of
$$ \frac{x(8x^2 + 2x + 3)}{(x^2 + 5x)}.$$

This would require us to use the product rule and the quotient rule. However SymPy has the built-in command `diff` which does exactly this for us.

In [32]:
expr_d = sp.diff(expr)
print(expr_d)

x*(-2*x - 5)*(8*x**2 + 2*x + 3)/(x**2 + 5*x)**2 + x*(16*x + 2)/(x**2 + 5*x) + (8*x**2 + 2*x + 3)/(x**2 + 5*x)


We can simplify this expression.

In [33]:
print(sp.simplify(expr_d))

(8*x**2 + 80*x + 7)/(x**2 + 10*x + 25)


The `subs` function is used to evaluate an expression at a given point.

In [34]:
a = expr_d.subs([(x, 5.5)])
print(a)

6.24943310657596


## Exercise

Use SymPy to evaluate the derivative of the function $f(x) = \tan^{-1}(\sin(x))$ at $x = \pi/12$. You will have to use the functions `sp.atan` and `sp.sin`.

In [63]:
# define x as a symbol
x = sp.symbols("x")

# create expression
f = ...

# differentiate f, use sp.diff(...)
df = ...

# evaluate df at x = pi/, use df.subs(...)
a = ...

# print results
pprint(a)
print(sp.simplify(a))

SymPy can also be used to solve equations using the `solve` command.

In [67]:
x1 = sp.solve(x**4 - 2*x**2, x)
print(x1)

[0, -sqrt(2), sqrt(2)]


The second argument "x" indicated the we are solving the expression for x. The return value is a list of all possible solutions.

## Exercise

Use the SymPy `solve` command to solve the quadratic equation 
$8x^2 + 16x + 4 = 0$.

Compare this to your answer from using the quadratic formula.

In [ ]:
x = sp.symbols("x")
expr = ...

x_sol = sp.solve(expr, x)

# print x_sol as floats

# NumPy

The `NumPy` module (http://www.numpy.org/) is an almost indespensible module for scientific computing. It provides objects such as arrays and matrices as well as functions spanning linear algebra, fourier transforms and stastics amoung numerous other things. 

Almost every assignment in this course will use NumPy. 

To start with we must import NumPy. To reduce the amount of typing for ourselves later we will call the module `np`.

In [70]:
import numpy as np

## Arrays

One important class that NumPy provides is the `array` class. An array is similar to a `list` in that it is a collection of objects. Typically arrays store numbers. 

NumPy arrays can be initialized in a similar way to lists.

In [71]:
a = np.array([1,2.0,3])
type(a)

numpy.ndarray

You'll notice the type of a is `numpy.ndarray`. NumPy arrays can be multidimensional. You can think of a 1D array as a kind of list (but not a Python list) and a 2D array as a kind of grid (or, if you know linear algebra as a matrix, but not actually a matrix). Higher dimensional arrays are certainly possible, you can think of a 3D array as a stack of grids. 

![np array](images/np_array.jpg)
(Image credit: Dalesha Hemrajani)


The attribute `ndim` stores the number of dimensions in the array.

In [72]:
a.ndim

1

Multidimensional arrays can be initialized as follows: 

In [79]:
b = np.array([[1, 2, 3.0], [1.2,2.2,2]])
print(b)

[[ 1.   2.   3. ]
 [ 1.2  2.2  2. ]]


In [80]:
print(b.ndim)

2


The `shape` property tells us how many rows and columns are in our array, while the `size` property tells us the total number of elements in the array.

In [78]:
print(b.shape)
print(b.size)

(2, 3)
6


Like lists and strings, arrays support indexing and slicing.

In [53]:
a[0]

1.0

In [54]:
b[1]

array([ 1.2,  2.2,  2. ])

In [55]:
b[1][0:2]

array([ 1.2,  2.2])

For those of you familiar with linear algebra, it may be tempting to think of 2D arrays as matrices. This is __not__ true. For example, if we call:

In [74]:
x = np.array([1,2])
A = np.array([[3,2], [1,2]])

print(A*x)

[[3 4]
 [1 4]]


We get a 2D array, instead of a vector. This is because array operations are defined elementwise. $A\mathbf{x}$ in this case is defined to be:

$$ \begin{bmatrix} A_{11}x_1 & A_{12}x_2\\ A_{21}x_1 & A_{22} x_2\end{bmatrix}$$

Likewise if we call $A^2$, we get

In [75]:
print(A**2)

[[9 4]
 [1 4]]


which is 
$$ \begin{bmatrix} A_{11}^2 & A_{12}^2\\A_{21}^2 & A_{22}^2\end{bmatrix}.$$
This is not what we expect from matrix squaring.

Fortunately NumPy has a matrix class that acts like we'd expect a matrix to act.

In [76]:
B = np.matrix([[3,2],[1,2]])
x = np.array([1,2])
B**2

matrix([[11, 10],
        [ 5,  6]])

To do matrix vector (or vector vector) products, we must use the `np.dot` command:

In [77]:
print(B.dot(x))
print(B.dot(x).ndim)

[[7 5]]
2


### Arrays vs. Lists

Arrays and lists are similar in many ways. Both represent a collection of objects. When should you use one over the other? 

For starters, arrays are mutable, however they do not support methods such as `pop` or `append`. Once initialized the size of an array cannot be easily changed. If your application needs to change the size of a collection, lists are the prefered option. 

In general lists are prefered unless there is a specific reason to use NumPy arrays. NumPy arrays are less flexible; however you'll find that they are needed to interact fully with the rest of NumPy. For example taking the dot product of two vectors requires the two vectors to be stored as NumPy arrays. 

It's possible to convert from a list to an array or vice versa. NumPy arrays provide the method `tolist()` which converts an array to a list.

In [56]:
c = a.tolist()
type(c)

list

NumPy also provides the function `asarray` that takes a list and returns an array.

In [57]:
d = np.asarray(c)
type(d)

numpy.ndarray